src/NHibernate/Transaction/AdoNetWithSystemTransactionFactory.cs

```diff
@@ -0,0 +1,384 @@
+									dependentSession.BeforeTransactionCompletion(null);
+							}
+						}
+						// Lock the session to ensure second phase gets done before the session is used by code following
+						// the transaction scope disposal.
+						Lock();
+
+						_logger.Debug("Prepared for system transaction");
+						preparingEnlistment.Prepared();
+					}
+					catch (Exception exception)
+					{
+						_logger.Error("System transaction prepare phase failed", exception);
+						preparingEnlistment.ForceRollback(exception);
+					}
+				}
+			}
+
+			void IEnlistmentNotification.Commit(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, true);
+
+			void IEnlistmentNotification.Rollback(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, false);
+
+			void IEnlistmentNotification.InDoubt(Enlistment enlistment)
+				=> ProcessSecondPhase(enlistment, null);
+
+			private void ProcessSecondPhase(Enlistment enlistment, bool? success)
+			{
+				using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+				{
+					_logger.Debug(
+						success.HasValue
+							? success.Value
+								? "Committing system transaction"
+								: "Rolled back system transaction"
+							: "System transaction is in doubt");
+					// we have not much to do here, since it is the actual
+					// DB connection that will commit/rollback the transaction
+					// After transaction actions are raised from TransactionCompleted event.
+
+					enlistment.Done();
+				}
+			}
+
+			#endregion
+
+			private void TransactionCompleted(object sender, TransactionEventArgs e)
+			{
+				try
+				{
+					e.Transaction.TransactionCompleted -= TransactionCompleted;
+					// This event may execute before second phase, so we cannot try to get the success from second phase.
+					// Using this event is required by example in case the prepare phase failed and called force rollback:
+					// no second phase would occur for this ressource. Maybe this may happen in some other circumstances
+					// too.
+					var wasSuccessful = false;
+					try
+					{
+						wasSuccessful =
+							AmbientTransaction.TransactionInformation.Status == TransactionStatus.Committed;
+					}
+					catch (ObjectDisposedException ode)
+					{
+						_logger.Warn("Completed transaction was disposed, assuming transaction rollback", ode);
+					}
+
+					// Allow transaction completed actions to run while others stay blocked.
+					_bypassWait.Value = true;
+					using (new SessionIdLoggingContext(_sessionImplementor.SessionId))
+					{
+						// Flag active as false before running actions, otherwise the session may not cleanup as much
+						// as possible.
+						IsInActiveTransaction = false;
+						// Never allows using connection on after transaction event.
+						using (_sessionImplementor.ConnectionManager.BeginsProcessingFromSystemTransaction(false))
+						{
+							_sessionImplementor.ConnectionManager.AfterTransaction();
```

Comment:
Now inside `using (_sessionImplementor.ConnectionManager.BeginsProcessingFromSystemTransaction(false))` for:

1. forbidding connection usage (`false` parameter), which by the way never causes a connection switch.
2. just flagging the connection as requiring a release, instead of releasing it from transaction completion event (NH-3023).